staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,615 @@
+// Otherwise the client should not execute the
+// request and afterExecution is irrelevant.
+func (qs *queueSetImpl) Wait(ctx context.Context, hashValue uint64) (tryAnother, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify locking is correct
+
+	qs.lockAndSyncTime()
+	// TODO(aaron-prindle) verify and test quiescing
+	// A call to Wait while the system is quiescing will be rebuffed by
+	// returning `tryAnother=true`.
+	if qs.emptyHandler != nil {
+		qs.lock.Unlock()
+		return true, false, nil
+	}
+
+	// ========================================================================
+	// Step 1:
+	// 1) Start with shuffle sharding, to pick a queue.
+	// 2) Reject old requests that have been waiting too long
+	// 3) Reject current request if there is not enough concurrency shares and
+	// we are at max queue length
+	// 4) If not rejected, create a packet and enqueue
+	req := qs.timeoutOldRequestsAndRejectOrEnqueue(hashValue)
+	// req == nil means that the request was rejected - no remaining
+	// concurrency shares and at max queue length already
+	if req == nil {
+		qs.lock.Unlock()
+		metrics.AddReject(qs.name, "queue-full")
+		return false, false, func() {}
+	}
+	// ========================================================================
+
+	// ------------------------------------------------------------------------
+	// Step 2:
+	// 1) The next step is to invoke the method that dequeues as much as possible.
+
+	// This method runs a loop, as long as there
+	// are non-empty queues and the number currently executing is less than the
+	// assured concurrency value.  The body of the loop uses the fair queuing
+	// technique to pick a queue, dequeue the request at the head of that
+	// queue, increment the count of the number executing, and send true to
+	// the request's channel.
+	qs.dequeueWithChannelAsMuchAsPossible()
+	// ------------------------------------------------------------------------
+	qs.lock.Unlock()
+
+	// ************************************************************************
+	// Step 3:
+	// After that method finishes its loop and returns, the final step in Wait
+	// is to `select` (wait) on a message from the enqueud packet's channel
+	// and return appropriately.
+	qs.counter.Add(-1)
+
+	select {
+	case execute := <-req.DequeueChannel:
+		if execute {
+			// execute the request
+			return false, true, func() {
+				qs.finishRequestAndDequeueWithChannelAsMuchAsPossible(req)
+			}
+
+		}
+		klog.V(5).Infof("request timed out after being enqueued\n")
+		metrics.AddReject(qs.name, "time-out")
+		return false, false, func() {}
+	case <-ctx.Done():
+		klog.V(5).Infof("request cancelled\n")
+		func() {
+			qs.lockAndSyncTime()
+			defer qs.lock.Unlock()
+
+			if req.Enqueued {
+				// remove the request from the queue as it has timed out
+
+				for i := range req.Queue.Requests {
+					if req == req.Queue.Requests[i] {
+						req.Queue.Requests = removeRequest(req.Queue.Requests, i)
+						break
+					}
+				}
```

Comment:
There is another thing that needs to be done if `req.Enqueued`.  By removing this request from its queue and _not_ starting execution, we might have just established the condition that requires a forked call on the EmptyHandler (if there is one).

Reply:
I've added the below (taken from the `Quiesce` implementation):
```
			if req.Enqueued {
				// remove the request from the queue as it has timed out
				for i := range req.Queue.Requests {
					if req == req.Queue.Requests[i] {
						req.Queue.Requests = removeRequest(req.Queue.Requests, i)
						break
					}
				}
				// At this point, if the qs is quiescing,
				// has zero requests executing, and has zero requests enqueued
				// then a call to the EmptyHandler should be forked.
				if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
					qs.getRequestsExecutingLocked() == 0 {
					qs.counter.Add(1)
					go func(eh fq.EmptyHandler) {
						eh.HandleEmpty()
						qs.counter.Add(-1)
					}(qs.emptyHandler)
				}
```

Reply:
good